src/libs/kata-sys-util/src/mount.rs

```diff
@@ -0,0 +1,1029 @@
+                device: fields[PROC_DEVICE_INDEX].to_string(),
+                path: fields[PROC_PATH_INDEX].to_string(),
+                fs_type: fields[PROC_TYPE_INDEX].to_string(),
+            });
+        }
+    }
+
+    Err(Error::NoMountEntry(mount_point.to_owned()))
+}
+
+/// Recursively create destination for a mount.
+///
+/// For a normal mount, the destination will always be a directory. For bind mount, the destination
+/// must be a directory if the source is a directory, otherwise the destination must be a normal
+/// file. If directories are created, their permissions are initialized to MountPerm.
+///
+/// # Safety
+///
+/// Every container has a root filesystems `rootfs`. When creating bind mounts for a container,
+/// the destination should always be within the container's `rootfs`. Otherwise it's a serious
+/// security flaw for container to read/override host side filesystem contents. Please refer to
+/// following CVEs for example:
+/// - [CVE-2021-30465](https://github.com/opencontainers/runc/security/advisories/GHSA-c3xm-pvg7-gh7r)
+///
+/// To ensure security, the `create_mount_destination()` function takes an extra parameter `root`,
+/// which is used to ensure that `dst` is within the specified directory. And a safe version of
+/// `PathBuf` is returned to avoid TOCTTOU type of flaws.
+pub fn create_mount_destination<S: AsRef<Path>, D: AsRef<Path>, R: AsRef<Path>>(
+    src: S,
+    dst: D,
+    _root: R,
+    fs_type: &str,
+) -> Result<impl AsRef<Path> + Debug> {
+    // TODO: enhance the implementation by `secure_path`.
+    let dst = dst.as_ref();
+    let parent = dst
+        .parent()
+        .ok_or_else(|| Error::InvalidPath(dst.to_path_buf()))?;
+    let mut builder = fs::DirBuilder::new();
+    builder.mode(MOUNT_PERM).recursive(true).create(parent)?;
+
+    if fs_type == "bind" {
+        // The source and destination for bind mounting must be the same type: file or directory.
+        if !src.as_ref().is_dir() {
+            fs::OpenOptions::new()
+                .mode(MOUNT_PERM)
+                .write(true)
+                .create(true)
+                .open(dst)?;
+            return Ok(dst.to_path_buf());
+        }
+    }
+
+    if let Err(e) = builder.create(dst) {
+        if e.kind() != std::io::ErrorKind::AlreadyExists {
+            return Err(e.into());
+        }
+    }
+    if !dst.is_dir() {
+        Err(Error::InvalidPath(dst.to_path_buf()))
+    } else {
+        Ok(dst.to_path_buf())
+    }
+}
+
+/// Remount a bind mount into readonly mode.
+///
+/// # Safety
+/// Caller needs to ensure safety of the `dst` to avoid possible file path based attacks.
+pub fn bind_remount_read_only<P: AsRef<Path>>(dst: P) -> Result<()> {
+    let dst = dst.as_ref();
+    if dst.is_empty() {
+        return Err(Error::NullMountPointPath);
+    }
+    let dst = dst
+        .canonicalize()
+        .map_err(|_e| Error::InvalidPath(dst.to_path_buf()))?;
+
+    do_remount_read_only(dst, MsFlags::empty())
```

Comment:
This function is specifically for bind mounts, but `do_remount_read_only()` is a general function fwics. Therefore, shouldn't this actually be:

```suggestion
    do_remount_read_only(dst, MsFlags::MS_BIND)
```

... and the bind flag removed from `do_remount_read_only()`?

Reply:
I would rename the function to `do_rebind_mount_read_only()`, otherwise it gets a little messy at line 348 below.